codec/enforceproto/extraneous_fields.go

```diff
@@ -0,0 +1,476 @@
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+
+	gogoproto "github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/protoc-gen-gogo/descriptor"
+	"google.golang.org/protobuf/encoding/protowire"
+
+	"github.com/cosmos/cosmos-sdk/codec/types"
+)
+
+const bit11NonCritical = 1 << 10
+
+type descriptorIface interface {
+	Descriptor() ([]byte, []int)
+}
+
+type protoMessageWithDescriptor struct {
+	descriptorIface
+	gogoproto.Message
+}
+
+var _ descriptor.Message = (*protoMessageWithDescriptor)(nil)
+
+type descriptorMatch struct {
+	cache map[int32]*descriptor.FieldDescriptorProto
+	desc  *descriptor.DescriptorProto
+}
+
+// CheckMismatchedProtoFields walks through the protobuf serialized bytes in b, and tries to
+// compare field numbers and wireTypes against what msg expects. The error returned if non-nil will contain
+// the listing of the extraneous fields by tagNumber and wireType, as well as mismatched
+// wireTypes.
+func CheckMismatchedProtoFields(b []byte, msg gogoproto.Message) error {
+	if len(b) == 0 {
+		return nil
+	}
+
+	desc, ok := msg.(descriptorIface)
+	if !ok {
+		return fmt.Errorf("%T does not have a Descriptor() method", msg)
+	}
+
+	rv := reflect.ValueOf(msg)
+	rt := rv.Type()
+
+	protoNumToGoTypesIndex, err := protoNumToGoType(rt)
+	if err != nil && err != errNoProtoTag {
+		return err
+	}
+
+	fieldDescProtoFromTagNum, _, err := descProtoCache(protoMessageWithDescriptor{
+		descriptorIface: desc,
+		Message:         msg,
+	})
+	if err != nil {
+		return err
+	}
+
+	for len(b) > 0 {
+		tagNum, wireType, n := protowire.ConsumeField(b)
+		if n < 0 {
+			return errors.New("invalid length")
+		}
+
+		fieldDescProto, ok := fieldDescProtoFromTagNum[int32(tagNum)]
+		switch {
+		case ok:
+			// Assert that the wireTypes match.
+			if !canEncodeType(wireType, fieldDescProto.GetType()) {
+				return &MismatchedWireType{
+					Type:         rt.String(),
+					TagNum:       tagNum,
+					GotWireType:  wireType,
+					WantWireType: protowire.Type(fieldDescProto.WireType()),
+				}
+			}
+
+		default:
+			if tagNum&bit11NonCritical == 0 {
+				// The tag is non-critical, so report it.
+				return &UnexpectedField{
+					Type:     rt.String(),
+					TagNum:   tagNum,
+					WireType: wireType,
+				}
+			}
+		}
+
+		// Given that the wireTypes match, now traverse that field.
+		// Find the field whose protoNumber was just looked up.
+		// Skip over the 2 bytes that store fieldNumber and wireType bytes.
+		fieldBytes := b[2:n]
+		b = b[n:]
+
+		// An unknown but critical field or just a scalar type (aka not a message).
+		if fieldDescProto == nil || fieldDescProto.IsScalar() {
+			continue
+		}
+
+		fieldGoType, ok := protoNumToGoTypesIndex[tagNum]
+
+		switch {
+		case fieldDescProto.OneofIndex != nil: // Oneof field.
+			// Traverse the oneof field, after locating it by its index!
+			oneOfs := extractOneOfTypes(rt)
+			oneOfSampleType := oneOfs[fieldDescProto.GetOneofIndex()]
+			fieldGoType := reflect.TypeOf(oneOfSampleType).Elem()
```

Comment:
Why do we need to use reflection at all at this point? We have the Descriptor, that should have all the information so we no longer need to introspect go structs at all at this point afaik.

Reply:
We also should be able to cache all that information in an index by pulling in referenced FileDescriptor's as needed based on the fields in the root message. That can be done in a single pass for a root proto.Message or whenever we encounter a type with an Any.

Reply:
>Why do we need to use reflection at all at this point? We have the Descriptor, that should have all the information so we no longer need to introspect go structs at all at this point afaik.

If you take a look at the generated proto, the wrappers aren’t
gogoproto.Message nor are there FieldDescriptorProto. The only way to walk
through their definitions is to use the OneofIndex, to get the definition
and then walk that struct, we have to us reflection at this point. I
confirmed that this is also what gogoproto does during Unmarshal.

>We also should be able to cache all that information in an index by pulling in referenced FileDescriptor's as needed based on the fields in the root message. That can be done in a single pass for a root proto.Message or whenever we encounter a type with an Any.

Yes we do in https://github.com/cosmos/cosmos-sdk/blob/bd488491ee5ff9faf87193c4e55af5512c48285f/codec/enforceproto/extraneous_fields.go#L448-L476